### PR TITLE
Chat with vision.

### PR DESCRIPTION
**Pull request by v9ai**

### CODE REVIEW

File: app/api/chat-with-vision/route.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import type { ChatWithVisionVariables } from "@/lib/types";
+
+const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
+const options: ClientOptions = {
+  apiKey: OPENAI_API_KEY,
+};
+
+export async function POST(req: Request) {
+  try {
+    const { imageURL, text } = (await req.json()) as ChatWithVisionVariables;
```
Comment (v9ai) [inline app/api/chat-with-vision/route.ts:21]: will need to figure out how to apply type properly and not cast

File: app/hooks/useChatWithVision.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+type Completions = OpenAI.Beta.Chat.Completions;
+
+const mutationFn = async (
+  args: ChatWithVisionVariables
+): Promise<Completions> => {
+  const { data } = await axios.post<Completions>("/api/chat-with-vision", args);
+  return data;
+};
+
+export const useChatWithVision = (): UseMutationResult<
```
Comment (v9ai) [inline app/hooks/useChatWithVision.ts:15]: in the consuming component I use like this

```
 const chatWithVision = useChatWithVision() ;
  .
  .
  .

{chatWithVision?.isPending ? <CircularProgress /> : <SendIcon />}
  ```

and I don't have to worry about storing loading state somewhere and resetting it

**Comment by athrael-soju** [comment]
Looks good, but best to have the complete feature before merging. Also, there's the upload button/route which are unused, so I wonder if we can make use of that route along with a message and sending them both when sending the message. This way we can also reuse the file upload with other types of extensions and if they aren't images, then we could USE RAG to answer questions for them. Or, we can use the assistant and in that case the file upload will instead add a file to an assistant/thread. So, tons of reusability for the upload route. We can even have api/upload/img, api/upload/doc and api/upload/assistant-doc

Thoughts?

**Comment by v9ai** [comment]
> Looks good, but best to have the complete feature before merging. Also, there's the upload button/route which are unused, so I wonder if we can make use of that route along with a message and sending them both when sending the message. This way we can also reuse the file upload with other types of extensions and if they aren't images, then we could USE RAG to answer questions for them. Or, we can use the assistant and in that case the file upload will instead add a file to an assistant/thread. So, tons of reusability for the upload route. We can even have api/upload/img, api/upload/doc and api/upload/assistant-doc
> 
> Thoughts?

wanted to test fast the experiment use-case, might go back and adjust as u mention

**Comment by athrael-soju** [comment]
@nicolad If we're not moving forward with this I'd suggest we close the PR and implement based on the current code.

**Pull request closed without merging**